lib/Runtime/ByteCode/ScopeInfo.cpp

```diff
@@ -162,23 +162,20 @@ namespace Js
         Scope* currentScope = byteCodeGenerator->GetCurrentScope();
         Assert(currentScope->GetFunc() == funcInfo);
 
-// Disable the below temporarily because of bad interaction with named function expressions
-#if 0
         if (funcInfo->root->IsDeclaredInParamScope()) {
             Assert(currentScope->GetScopeType() == ScopeType_FunctionBody);
             Assert(currentScope->GetEnclosingScope());
 
-            FuncInfo* func = currentScope->GetEnclosingScope()->GetFunc();
+            FuncInfo* func = byteCodeGenerator->GetEnclosingFuncInfo();
             Assert(func);
 
-            if (func->IsBodyAndParamScopeMerged())
+            if (func->IsBodyAndParamScopeMerged() && !(func->funcExprScope && func->funcExprScope->GetCanMerge()))
```

Comment:
With the fix to get the right enclosing func, do we still need this change to the "if" condition? Was the problem with the named function expression case just the fact that it prevented us from getting the right enclosing func?

Reply:
Yeah, I tried removing the check for the func expr scope but we still need it. The problem is that the function declared in the param scope needs to have access to the body scope of it's enclosing function when the function expression scope is merged with the body and param scope.

Reply:
Are you sure? A function declared in the param scope shouldn't have access to the declaring function's body scope. I thought that was the original bug. Is it rather that it needs to have access to the name of the declaring function?

Reply:
Yes, that's the issue. The name of the enclosing function is not in the param scope. It's in the funcexpr scope, though, from the looks of it.

Reply:
Okay. So, regardless, the scope info should be recorded from the declaring function's param scope outward, including (if necessary) the declaring function's function expression scope, but not the declaring function's body scope. So that all seems right. If references to the declaring function's name are not getting bound correctly, I wonder if there's something wrong with the recording of scope info for that case?